### DISCUSSION TITLE
Drafting a Docker Swarm stack for high availability and performance

### DISCUSSION THREAD

**Discussion by cycneuramus** [seaweedfs/seaweedfs#4105] [Show and tell]
Hi all,

I thought I might share [this](https://github.com/cycneuramus/seaweedfs-docker-swarm) draft of an example Docker Swarm stack for achieving a performant and highly available SeaweedFS setup ~(with some help from JuiceFS)~ _[EDIT: redesigned in accordance with the discussion below]_. The architecture of this particular stack might prove useful to those who, like me, are running geographically distributed nodes in Swarm and want to avoid performance issues related to network latency.

There are some hacks involved, notably the use of Docker-in-Docker for the [global mounts](https://github.com/cycneuramus/seaweedfs-docker-swarm/blob/master/docker-compose.yml#L60-L79), but my initial tests are quite promising nonetheless. Still, use with caution.

To those who are interested, feel free to suggest improvements or point out any misunderstandings I might have about the various configuration parameters and so on.

**Comment by chrislusf** [comment]
Please try with weed mount after 3.38. It should be much more solid and can simplify the architecture.

**Comment by cycneuramus** [comment]
Thanks for the tip! About the architecture, do you mean that instead of running a KeyDB instance + JuiceFS mount on every node (as I do to reduce latency between mount and metadata store), I run a filer + weed mount on every node instead?

**Comment by cycneuramus** [comment]
Weed mount seems to perform about 10x slower than my initial architecture. Starting a file transfer to the mount—the same 10 GB folder I used to test the JuiceFS mounts—the speed is comparable initially (~30–50 MiB/s) but quickly drops down to about 3 MiB/s and stays there. Is that perhaps to do with the replication between filers/mounts?

Running 1–2 filers only instead of one per node made no difference here either, but that was probably to be expected given the network latency between my VPN connected nodes (where e.g. weed mount on node1 may be writing to filer on node2).

**Reply by chrislusf** [reply]
please make a reproducible case

**Reply by chrislusf** [reply]
Remote shared filer is fine. No need to setup local filers. It only adds to local CPU cost.

**Reply by cycneuramus** [reply]
Thanks, that's good to know. Just to confirm, this is true even when there is network latency between mount (on node A in country X) <—> filer (on node B in country Y), i. e. a filer on node A would not make for better performance here?

In any case, my observations on weed mount performance were only the result of some very quick testing, basically just switching out the JuiceFS mounts with weed mounts and standalone keydb with embedded leveldb on the filers (and BTW I tested both with one filer on every node and 1–2 filers total). 

So it may not be worth your time to troubleshoot with me at this stage. Nevertheless, I'll try to find some time to get back to you with an attempt at a reproducible case. In the meantime, thanks for your willingness to help and for this awesome piece of free software!

**Reply by chrislusf** [reply]
the latency matters when committing any metadata to filer.

I would like to know what is your usage pattern. Is it randomly read write a lot of small/large files, or a few large files. Some pseudo code would be helpful.

**Reply by cycneuramus** [reply]
I created a [separate branch](https://github.com/cycneuramus/seaweedfs-juicefs-docker/tree/seaweedfs-only) of my repo that demonstrates the SeaweedFS-only setup I used to test out weed mount. The crucial thing, and one that I can't really include in a reproducible case, is that all Docker Swarm nodes are geographically distributed and connected via Wireguard; hence my concern with network latency.

As for my usage pattern, it's definitely on the side of randomly reading and writing predominantly small files. More specifically, I'm toying with using SeaweedFS as the persistent storage for various Docker Swarm services, hence the files consist of whatever is output by those services, including database files (yes, I know, I'm only testing!), jpegs, logs, and so on.

In my testing, I transfered the 13 GB directory containing all of these files to the weed / juicefs mount using the following command (`rclone` local-to-local):

``` 
rclone sync \
    /path/to/src-dir \
    /path/to/mountpoint \
    --create-empty-src-dirs \
    --links \
    --transfers=6 \
    --progress
```

**Comment by chrislusf** [comment]
My local test is fine. I wonder how many files you are actually testing.
```
 > rclone sync . ~/tmp/mm/vm/ --create-empty-src-dirs \
    --links \
    --transfers=6 \
    --progress
Transferred:   	    6.104 GiB / 6.104 GiB, 100%, 114.885 MiB/s, ETA 0s
Transferred:          497 / 497, 100%
Elapsed time:        44.3s
```

**Reply by cycneuramus** [reply]
```
$ find /my/test/dir -type f | wc -l
90479
```

Like I intimated before, though, my quick testing here is probably not very representative given my oddball network topology in combination with replication settings (`200`), etc. 

Also—and crucially!—there was some unintentional cheating going on with my initial architecture since I had enabled the [`--writeback`](https://juicefs.com/docs/community/cache_management/#writeback) feature of JuiceFS which, in my case, created an illusion of fast transfer speeds. Without that flag, it seems the actual transfer speed is comparable to weed mount.

So I think you can totally forget about my reports on the performance differences. My actual problem, rather, is figuring out how to optimize for performance in a Wireguard mesh of geographically distributed servers. I'll keep tinkering and see what I can find.

**Comment by cycneuramus** [comment]
UPDATE: 

I have been experimenting with different configuration parameters and seem to have landed on a SeaweedFS-only setup that makes for way better write speeds than what I initially reported with weed mount. An important factor here was changing the default replication setting from `200` to `000` and, as a workaround, implementing a [cron-based maintenance schedule](https://github.com/cycneuramus/seaweedfs-juicefs-docker/blob/169e86c0cd5581824e81e906df258b35980e5e22/docker-compose.yml#L81-L104) to replicate volumes "manually" once a day. Not quite the seamless HA setup that I was going for initially, but the tradeoff is totally acceptable for my use-case. 

My current setup can be perused [here](https://github.com/cycneuramus/seaweedfs-juicefs-docker/tree/seaweedfs-only). I'll probably make this the master branch and ditch the entire JuiceFS architecture (and also edit the opening post here + repo readme once I've done some further testing). If you or anyone else is interested, and have some pointers on how further to improve performance in this kind of mesh-based network topology, please let me know. Anyway, thanks for nudging me in the right direction!

As it stands right now, there is one issue remaining. In certain circumstances, read performance seems to be slow as molasses. For instance, I am self-hosting a Gitea instance, and if I run it from local storage, the page load time when browsing to my repo (i.e. https://git.mydomain.com/myuser/myrepo) is < 1 second. When running the same instance from my SeaweedFS mountpoint, however, the same page load may take up to 10 full seconds. This kind of slowdown does not occur in every case; Vaultwarden, for example, remains quite snappy when running from the FUSE mount, whereas Nextcloud exhibits the same slow behaviour as Gitea. So this probably depends on what the apps are doing behind the scenes to load their content. 

I'm mounting with a generous `cacheCapacityMB` of `15000`, but this does not seem to help. Do you have any idea how to troubleshoot this? I could provide some Docker Compose files if necessary.

**Reply by chrislusf** [reply]
The caching algo would need some tuning for special access patterns. Need a way to reproduce the problem.

**Reply by cycneuramus** [reply]
Gotcha. Don't know how much it helps, but here is a docker compose setup to launch an instance of Gitea at http://localhost:3000:

```
version: "3.9"

services:
  app:
    image: gitea/gitea:latest-rootless
    user: 1000:1000
    ports:
      - 3000:3000
    environment:
      - USER_UID=1000
      - USER_GID=1000
      - GITEA__database__DB_TYPE=postgres
      - GITEA__database__HOST=db:5432
      - GITEA__database__NAME=gitea
      - GITEA__database__PASSWD=gitea
      - GITEA__database__USER=gitea
      - GITEA__server__ROOT_URL=http://localhost:3000
    volumes:
      - /path/to/fusemount/data:/var/lib/gitea
      - /path/to/fusemount/config:/etc/gitea
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro

  db:
    image: postgres:15-alpine
    environment:
      - POSTGRES_DB=gitea
      - POSTGRES_USER=gitea
      - POSTGRES_PASSWORD=gitea
      - PG_DATA=/var/lib/postgresql/data
    user: 1000:1000
    volumes:
      - /path/to/fusemount/db:/var/lib/postgresql/data
      - /etc/passwd:/etc/passwd:ro
```

Bring up the container: 

`mkdir /path/to/fusemount/{data,config,db}`
`docker-compose up -d`

Browse to http://localhost:3000 to finish install. Then create a user:

`docker exec -it <gitea app container> gitea admin user create --username myname --password asecurepassword --email me@example.com`

... and push a repo (the one I'm testing against is about 4 mb in size).

**Reply by cycneuramus** [reply]
For posterity and potential troubleshooters of the future: 

The issue with Gitea in particular can be addressed by configuring a more aggressive app cache, e. g. by spinning up a Redis instance and adding to [`app.ini`](https://docs.gitea.io/en-us/config-cheat-sheet/):

```
[cache]
ENABLED = true
ADAPTER = redis
HOST = redis://<redis_host>:<redis_port>/0?pool_size=100&idle_timeout=180s

[cache.last_commit]
ENABLED=true
COMMITS_COUNT=50
```

Nextcloud—even with Redis and other optimizations—seems to be a different beast.

**Comment by legout** [comment]
Hi, this is a really interesting discussion.

@cycneuramus Do you still run this setup? If so, what are the lessons learned from this after nearly two years?

Thanks
legout
